gix-archive/src/archive.rs

```diff
@@ -0,0 +1,44 @@
+use std::io::{Read, Write};
+use std::path::PathBuf;
+use std::slice::Iter;
+
+#[derive(Debug, thiserror::Error)]
+#[allow(missing_docs)]
+pub enum Error {
+    #[error("Failed to archive")]
+    FailedToArchive,
+}
+
+/// File to be added to the archive.
+/// It is processed (contents after applying .gitattributes/filtering)
+/// TODO Do we need to care about empty dirs and links?
+pub struct File<R: Read> {
+    pub path: PathBuf,
+    pub data: R,
+}
+
+/// Available formats
+#[derive(PartialEq)]
+pub enum Format {
```

Comment:
I think with a scaling implementation and support for more customization there will be a trait of sorts to be implement by each archiving backend. However, I can also imagine that the selection of these backends could be hidden and an `enum` allows to choose them easily (along with feature toggles to pull in their dependencies on demand).

Reply:
Added `Custom(FromConfig)` variant

Reply:
I am biased towards using a trait for this, but in theory all that can be done differently too and I am open to that. This is exploration so ideally there is tests for the anticipated usecases.